File: test/Projection.py
import unittest as ut
import numpy as np
import Maniverse as mv
from scipy.linalg import expm

# Orthogonal projection
# Finding the Stiefel matrix closest to the given matrix A
# Minimize L(C) = || C - A ||^2
# A \in R(10, 6)
# C \in St(10, 6)

class Obj(mv.Objective):
	def __init__(self):
		super().__init__()
		self.A = np.loadtxt("Sym10.txt", delimiter = ',')[:60].reshape([6, 10]).T

	def Calculate(self, C, derivatives):
		if 0 in derivatives:
			self.Value = np.linalg.norm(C[0] - self.A) ** 2
		if 1 in derivatives:
			self.Gradient = [ 2 * ( C[0] - self.A ) ]

	def Hessian(self, X):
		return [ 2 * X[0] ]

class AndersonObj(Obj):
	def Calculate(self, C, derivatives):
		super().Calculate(C, derivatives)
		if 1 in derivatives:
			self.Gradient = [ -2 * ( C[0] - self.A ) ]

class TestProjection(ut.TestCase):
	def __init__(self, *args):
		super().__init__(*args)
		self.Obj = Obj()
		self.AndersonObj = AndersonObj()
		U, _, Vt = np.linalg.svd(self.Obj.A, full_matrices = False)
		self.Manifold = mv.Stiefel( U @ Vt @ expm( self.Obj.A[4:, :] - self.Obj.A[4:, :].T ) )
		self.Solution = U @ Vt
		self.Tolerance = (1.e-5, 1.e-5, 1.e-5)

	def testNewtonCG(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		tr = mv.TrustRegion()
		cg = mv.ConjugateGradient(M, 0, 1, (1e-4, 1e-4), M.getDimension(), 0)
		converged = mv.Newton(
				M, tr, cg, self.Tolerance, 9, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P, self.Solution, atol = 1e-5)

	def testLBFGS(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		converged = mv.LBFGS(
				M, self.Tolerance,
				20, 19, 0.1, 0.75, 5, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P, self.Solution, atol = 1e-5)

	def testAnderson(self):
		M = mv.Iterate(self.AndersonObj, [self.Manifold])
		converged = mv.Anderson(
				M, self.Tolerance,
				0.2, 6, 28, 0
		)
		assert converged
		assert np.allclose(M.Ms[0].P, self.Solution, atol = 1e-5)

	def testLanczos(self):
		M = mv.Iterate(self.Obj, [self.Manifold])
		M.setPoint([self.Solution], 1)
		M.Func.Calculate(M.getPoint(), [0, 1, 2])
		M.setGradient()
		Evals, Evecs = mv.Lanczos(M, M.getDimension(), 0, 0, 0)
		for i in range(len(Evecs)):
			residual = np.linalg.norm( M.ConstraintProjectedHessian(Evecs[i]) - Evals[i] * Evecs[i] )
			assert residual < 1e-5

if __name__ == "__main__":
	TestProjection().testNewtonCG()
	TestProjection().testLBFGS()
	TestProjection().testAnderson()
	TestProjection().testLanczos()

File: test/Projection.cpp
#include <Eigen/Dense>
#include <unsupported/Eigen/MatrixFunctions>
#include <cmath>
#include <string>
#include <vector>
#include <memory>
#include <iostream>
#include <Maniverse/Manifold/Stiefel.h>
#include <Maniverse/Optimizer/TruncatedNewton.h>
#include <Maniverse/Optimizer/LBFGS.h>
#include <Maniverse/Optimizer/Anderson.h>

// Orthogonal projection
// Finding the Stiefel matrix closest to the given matrix A
// Minimize L(C) = || C - A ||^2
// A \in R(10, 6)
// C \in St(10, 6)

namespace mv = Maniverse;

class ObjProjection: public mv::Objective{ public:
	Eigen::MatrixXd A = Eigen::MatrixXd::Zero(10, 6);

	ObjProjection(){
		const double data[] = {
			#include "Sym10.txt"
		};
		std::memcpy(A.data(), &data, 10 * 6 * 8);
	};

	virtual void Calculate(std::vector<Eigen::MatrixXd> C, int /*derivative*/) override{
		Value = ( C[0] - A ).squaredNorm();
		Gradient = { 2 * ( C[0] - A ) };
	};

	std::vector<Eigen::MatrixXd> Hessian(std::vector<Eigen::MatrixXd> V) const override{
		return std::vector<Eigen::MatrixXd>{ 2 * V[0] };
	};
};

class AndersonObjProjection: public ObjProjection{ public:
	void Calculate(std::vector<Eigen::MatrixXd> C, int /*derivative*/) override{
		ObjProjection::Calculate(C, 0);
		Gradient = { - 2 * ( C[0] - A ) };
	};
};

#define __Check_Result__\
	std::cout << typeid(*this).name() << " " << __func__ << " ";\
	if ( converged ){\
		if ( ( M.Ms[0]->P - Solution ).cwiseAbs().maxCoeff() < 1e-5 ){\
			std::cout << "\033[32mSuccess!\033[0m" << std::endl;\
		}else std::cout << "\033[31mFailed: Incorrect solution!\033[0m" << std::endl;\
	}else std::cout << "\033[31mFailed: Not converged!\033[0m" << std::endl;

class TestProjection{ public:
	ObjProjection Obj = ObjProjection();
	AndersonObjProjection AndersonObj = AndersonObjProjection();
	mv::Stiefel Manifold = mv::Stiefel(Eigen::MatrixXd::Identity(10, 6));
	std::tuple<double, double, double> Tolerance = {1.e-5, 1.e-5, 1.e-5};
	mv::TrustRegion TrustRegion = mv::TrustRegion();
	Eigen::MatrixXd Solution = Eigen::MatrixXd::Identity(10, 6);

	TestProjection(){
		Eigen::BDCSVD<Eigen::MatrixXd, Eigen::ComputeThinU | Eigen::ComputeThinV> svd(Obj.A);
		const Eigen::MatrixXd U = svd.matrixU();
		const Eigen::MatrixXd V = svd.matrixV();
		Manifold = mv::Stiefel( U * V.transpose() * ( Obj.A.bottomRows(6) - Obj.A.bottomRows(6).transpose() ).exp() );
		Solution = U * V.transpose();
	};

	void testTruncatedNewton(){
		mv::Iterate M(Obj, {Manifold.Share()}, true);
		const bool converged = mv::TruncatedNewton(
				M, TrustRegion, Tolerance,
				0.001, 9, 1
		);
		__Check_Result__
	};

	void testLBFGS(){
		mv::Iterate M(Obj, {Manifold.Share()}, true);
		const bool converged = mv::LBFGS(
				M, Tolerance,
				20, 19, 0.1, 0.75, 5, 1
		);
		__Check_Result__
	};

	void testAnderson(){
		mv::Iterate M(AndersonObj, {Manifold.Share()}, true);
		const bool converged = mv::Anderson(
				M, Tolerance,
				0.2, 6, 28, 1
		);
		__Check_Result__
	};
};

int main(){
	TestProjection().testTruncatedNewton();
	TestProjection().testLBFGS();
	TestProjection().testAnderson();
}
